pkg/component/controller/tunneledapiendpointreconciller.go

```diff
@@ -0,0 +1,214 @@
+}
+
+func (ep TunneledEndpointReconciler) Init() error {
+	return nil
+}
+
+func (ep *TunneledEndpointReconciler) Run(ctx context.Context) error {
+	go func() {
+		ticker := time.NewTicker(10 * time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				err := ep.reconcile(ctx)
+				if err != nil {
+					ep.logger.Warnf("external API address reconciliation failed: %s", err.Error())
+				}
+			case <-ctx.Done():
+				ep.logger.Info("endpoint reconciler done")
+				return
+			}
+		}
+	}()
+	return nil
+}
+
+func (ep *TunneledEndpointReconciler) Stop() error {
+	return nil
+}
+
+func (ep *TunneledEndpointReconciler) Healthy() error {
+	return nil
+}
+
+func (ep *TunneledEndpointReconciler) Reconcile(ctx context.Context, cfg *v1beta1.ClusterConfig) error {
+	ep.cfg = cfg
+	return nil
+}
+
+func (ep TunneledEndpointReconciler) reconcile(ctx context.Context) error {
+	if ep.cfg == nil {
+		return nil
+	}
+	if !ep.leaderElector.IsLeader() {
+		ep.logger.Debug("we're not the leader, not reconciling api endpoints")
+		return nil
+	}
+
+	if !ep.cfg.Spec.API.TunneledNetworkingMode {
+		return fmt.Errorf("impossible to disable tunneled networking for the KAS server on the fly, restart the node")
+	}
+
+	if err := ep.makeDefaultServiceInternalOnly(ctx); err != nil {
+		return fmt.Errorf("can't make `kubernetes` service be internal only: %w", err)
+	}
+
+	if err := ep.reconcileEndpoint(ctx); err != nil {
+		return fmt.Errorf("can't reconcile endpoint for the default service: %w", err)
+	}
+	return nil
+}
+
+func (ep TunneledEndpointReconciler) reconcileEndpoint(ctx context.Context) error {
+	c, err := ep.kubeClientFactory.GetClient()
+	if err != nil {
+		return err
+	}
+
+	epClient := c.CoreV1().Endpoints("default")
+
+	addresses, err := makeNodesAddresses(ctx, c)
+	if err != nil {
+		return err
+	}
+	if len(addresses) == 0 {
+		return nil
+	}
+	subsets := []v1core.EndpointSubset{
+		v1core.EndpointSubset{
+			Addresses: addresses,
+			Ports: []v1core.EndpointPort{
+				v1core.EndpointPort{
+					Name:     "https",
+					Protocol: "TCP",
+					Port:     6443,
+				},
+			},
+		},
+	}
+	kubernetesEndpoint, err := epClient.Get(ctx, "kubernetes", v1.GetOptions{})
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return ep.createEndpoint(ctx, subsets)
+		}
+		return err
+	}
+
+	kubernetesEndpoint.Subsets = subsets
+	_, err = epClient.Update(ctx, kubernetesEndpoint, v1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func makeNodesAddresses(ctx context.Context, c kubernetes.Interface) ([]v1core.EndpointAddress, error) {
+	nodes, err := c.CoreV1().Nodes().List(ctx, v1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("can't list nodes: %w", err)
+	}
+
+	addresses := make([]v1core.EndpointAddress, 0, len(nodes.Items))
+	var ipAddr string
+	for _, node := range nodes.Items {
+		node := node
+		ipAddr = ""
+		for _, addr := range node.Status.Addresses {
+			if addr.Type != v1core.NodeInternalIP {
```

Comment:
We need to verify if there are cases when node might have ONLY public address set. Those cases might be the ones on a cloud where node has only public address and the cloud-controller sets that as public on the node objet.

Reply:
I wonder what should we do in such case - use the public address, skip until internal address would be there, or just use external address from config (basically dropping the tunneling). Do you have any opinion?

Reply:
if there's only public address set for node, I _think_ those are only in case the node actually has that address in its interface. So no NAT etc. Which makes me think we can use the address as-is.

So basically we'd take the `InternalAddress` and if node does not have one, then `ExternalAddress`